lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -974,6 +1047,9 @@ void Inliner::run() {
       if (auto ref = p.dyn_cast<InnerRefAttr>())
         instOpHierPaths[ref].push_back(nla.sym_nameAttr());
   }
+  // Mark 'module-only' the NLA's that only target modules.
+  // These may be deleted when their module is inlined/flattened.
+  identifyNLAsTargetingOnlyModules();
```

Comment:
I was wondering if this preprocess step is really necessary , instead of doing this lazily ?
The `Inliner::cloneAndRename` will eventually visit every nonlocal annotation specified on an operation inside the module, that is inlined/flattened.  Maybe we can mark an NLA dead/live after a module is cloned ?
If this doesn't affect the performance much, then it maybe totally fine to just do this.

Reply:
I considered that, but since the code [is careful to update the (mutable)NLA's _before_ visiting operations](https://github.com/llvm/circt/blob/31511a79f153c7229ce7121ec0c41ffa03229da0/lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp#L914) (search for "Preorder" for other locations), I wasn't sure this approach would work (or that it'd be obviously correct if it did).  It does appear that the updating before cloning is more for the local/non-local handling, not dead/alive.

Thinking out loud a little...
As a small addition (that you likely already were thinking of), I think your suggested approach would also need to add something to `mapPortsToWires` so those keep their NLA's alive as well (for ports too not just operations visited in `cloneAndRename`), be that setting a flag on the MNLA ("referenced by a non-module target") for checking in a later GC-ish processing, and/or maintaining a set of "live" NLA's or mutable NLA's or however that would work (hmm, how would the Inliner know which NLA's to go through and check whether they are now dead ("mark an NLA dead/live after a module is cloned"), at the point after cloning the ops/ports-- re-walking to find them again ofc largely defeats the purpose of avoiding the pre-process, modulo locality I suppose; if we can defer determining dead-ness until the end then that simplifies things).   Not saying this isn't resolvable, just some details to sort I suppose :smile:.  Maybe this is simpler than I'm thinking :innocent: , my headspace is a bit full of other details at the moment.

I suspect we might be able to do as you say re:moving dead/live handling afterwards (and generally avoiding the preprocess step).  That does seem potentially more elegant, assuming it doesn't add much more complexity / tracking to things.

I can take a look soon (TM), or do you have a plan in mind? I know you've been working with this code and NLA's quite a bit.  We can chat offline about it too if you'd like! :+1: 

In the meantime I'll look at this some more/think a bit and go from there.  Thanks!

Reply:
Okay!

Also, performance on large design seems to at least not be problematic, before/after timings producing rather overlapping distributions with maybe this shifting average up a small amount.

Finding a more integrated solution would be nice, thanks!